ORC-250 - Create sha256 mask

moresandeep opened the pull request:
Masking strategy that masks String, Varchar, Char and Binary types
as SHA 256 hash.

**For String type:**
All string type of any length will be converted to 64 length SHA256 hash.

**For Varchar type:**
For Varchar type, max-length property will be honored i.e.
if the length is less than max-length then the SHA256 hash will be truncated
to max-length. If max-length is greater than 64 then the output is the sha256
length, which is 64.

**For Char type:**
For Char type, the length of mask will always be equal to specified max-length.
If the given length (max-length) is less than SHA256 hash length (64)
the mask will be truncated.
If the given length (max-length) is greater than SHA256 hash length (64)
then the mask will be padded by blank spaces.

**For Binary type:**
All Binary type of any length will be converted to 64 length SHA256 hash.

omalley commented on java/core/src/java/org/apache/orc/impl/mask/SHA256MaskFactory.java:
```diff
@@ -0,0 +1,290 @@
+    }
+
+    /* For type char */
+    if (schema.getCategory() == TypeDescription.Category.CHAR) {
```
Probably a switch would be clearer here.

moresandeep replied:
Sure, I'll use a Switch statement here.

omalley commented on java/core/src/java/org/apache/orc/impl/mask/SHA256MaskFactory.java:
```diff
@@ -0,0 +1,290 @@
+        .wrap(source.vector[row], source.start[row], source.length[row]);
+
+    // take SHA-256 Hash and convert to HEX
+    byte[] hash = DatatypeConverter
```
Let's put the hash back as bytes rather than hex digits for the binary types.

omalley commented on java/core/src/java/org/apache/orc/impl/mask/SHA256MaskFactory.java:
```diff
@@ -0,0 +1,290 @@
+
+    if (targetLength > hash.length) {
+
+      System.arraycopy(hash, 0, outputBuffer, 0, hash.length);
```
You don't need to copy the output bytes, just assign the hash variable into the target by setting:
* target.vector[r] = hash
* target.start[r] = 0
* target.length[r] = targetLength

moresandeep commented:
Hello @omalley 
Thank you for the review !
I have incorporated most of the changes you proposed, except the one about copying output bytes.
The thing is I will be needing another buffer for case where the targetLength > hash.length to pad the hash with blanks. So, I am wondering if we can keep that one as it is, I will do this in the maskBinary function where I do not need to pad. Let me know your thoughts. 
Also I have added hardened some test cases.
Thanks again.

moresandeep commented:
Updated the PR with suggested changes.
